File: taskvine/src/examples/vine_example_watch.py
#!/usr/bin/env python

# This example program shows the behavior of the watch parameter when adding an
# output file to a task.

# If a task produces output to a file incrementally as it runs,
# it can be helpful to see that output piece by piece as it
# is produced. By simply adding watch=True to the output
# of the program, taskvine will periodically check for output
# and return it to the manager while each task runs.  When the
# task completes, any remaining output is fetched.

# This example runs several instances of the task named
# trickle.sh, which gradually produces output
# every few seconds.  While running the manager program, open
# up another terminal, and observe that files output.0, output.1,
# etc are gradually produced throughout the run.

import ndcctools.taskvine as vine
import sys

script = """
#!/bin/sh
# This is a simple example of a program that gradually
# produces output over time.  It just logs the current
# time every second for 30 seconds.

hostname 

for n in $(seq 1 30)
do
	sleep 1 
	date
done

echo "done!"
""";


if __name__ == "__main__":
    m = vine.Manager()
    print("listening on port", m.port)

    script = m.declare_buffer(script,cache=True)
    
    n = 3
    for i in range(n):
        t = vine.Task("./trickle.sh > output")

        t.add_input(script, "trickle.sh")

        output = m.declare_file(f"output.{i}")
        t.add_output(output, "output", watch=True)

        t.set_cores(1)

        m.submit(t)

    print("Waiting for tasks to complete...")

    while not m.empty():
        t = m.wait(5)
        if t:
            if t.successful():
                print(f"task {t.id} result: {t.std_output}")
            elif t.completed():
                print(f"task {t.id} completed with an executin error, exit code {t.exit_code}")
            else:
                print(f"task {t.id} failed with status {t.result}")

        # print to the console the contents of the files being watched
        for i in range(n):
            try:
                with open(f"output.{i}") as f:
                    print(f"output.{i}:\n{f.readlines()}\n")
            except IOError:
                pass

    print("All tasks complete!")
# vim: set sts=4 sw=4 ts=4 expandtab ft=python:

File: taskvine/src/examples/vine_example_watch.c
/*
This example program shows the behavior of the VINE_WATCH flag.

If a task produces output to a file incrementally as it runs,
it can be helpful to see that output piece by piece as it
is produced. By simply adding the VINE_WATCH flag to the output
of the program, taskvine will periodically check for output
and return it to the manager while each task runs.  When the
task completes, any remaining output is fetched.

This example runs several instances of the task named
trickle.sh which gradually produces output
every few seconds.  While running the manager program, open
up another terminal, and observe that files output.0, output.1,
etc are gradually produced throughout the run.
*/

#include "taskvine.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

const char *script = "#!/bin/sh\nhostname\nfor n in $(seq 1 30)\ndo\nsleep 1\ndate\ndone\necho \"done!\"\n";

int main(int argc, char *argv[])
{
	struct vine_manager *m;
	struct vine_task *t;

	m = vine_create(VINE_DEFAULT_PORT);
	if(!m) {
		printf("Couldn't create manager: %s\n", strerror(errno));
		return 1;
	}
	printf("Listening on port %d...\n", vine_port(m));

	struct vine_file *scriptfile = vine_declare_buffer(m,script,strlen(script), VINE_CACHE);

	int i;
	for(i=0;i<10;i++) {
		char output_name[256];
		sprintf(output_name,"output.%d",i);
		struct vine_file *output_file = vine_declare_file(m, output_name, VINE_CACHE);

		t = vine_task_create("./trickle.sh > output");
		vine_task_add_input(t, scriptfile, "trickle.sh", 0);
		vine_task_add_output(t, output_file, "output", VINE_WATCH );
		vine_task_set_cores(t,1);
		vine_submit(m, t);
	}

	printf("Waiting for tasks to complete...\n");

	while(!vine_empty(m)) {
		t = vine_wait(m, 5);
		if(t) {
			vine_result_t r = vine_task_get_result(t);
                        int id = vine_task_get_id(t);

			if(r==VINE_RESULT_SUCCESS) {
				printf("Task %d complete: %s\n",id,vine_task_get_command(t));
                        } else {
                                printf("Task %d failed: %s\n",id,vine_result_string(r));
                        }

                        vine_task_delete(t);
		}
	}

	printf("All tasks complete!\n");

	vine_delete(m);

	return 0;
}
